apps/prairielearn/src/lib/editors.ts

```diff
@@ -231,9 +236,18 @@ export abstract class Editor {
           );
         };
 
+        let sharingSyncError = false;
         try {
           await writeAndCommitChanges();
-
+          const result = await syncFromDisk.validateCourseData(
+            this.course.id,
+            this.course.path,
+            logger,
+          );
```

Comment:
This will do a full load+validate of course data from disk. This is IMO unacceptably slow to do as part of the file editing process. For large courses like Mariana's, this process alone takes more than 500ms, which would then be a strict lower bound on how fast we could make syncing.

From looking at the implementation of `validateCourseData`, the only thing you really need is the list of questions in the DB and the list of questions on disk. You don't care about any other course data, and you also don't care about the contents of any of the question JSON files. I'd be inclined to optimize this function as much as possible such that it only computes and loads what is absolutely necessary, which may necessitate bespoke code. I'm not sure if this is the most optimized way to implement this, but one option would be to first query for the list of shared questions and then assert that the JSON file for each one exists on disk. This makes the complexity of the operation scale with the number of shared questions in a course, not the total number of questions or anything else.

Somewhat related: the name `validateCourseData` seems misleading it doesn't validate course data as a whole, it very specifically validates whether any shared questions were deleted or renamed.

Reply:
resolved as we discussed in our meeting, by keeping courseData around whenever possible. 

Now, the only case that has a performance hit is if someone has edited in git, and then use the browser file editor right after that.

Reply:
I also named the function now `validateSharingConfiguration` rather than the misleading `validateCourseData`